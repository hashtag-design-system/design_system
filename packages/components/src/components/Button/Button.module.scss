.button {
  color: blue;
  display: inline-flex;
  gap: 40px;
  align-items: center;
  padding-left: 20px;
  padding-right: 20px;

  h1 {
    order: 2;
  }

  p {
    order: 1;
  }
}
