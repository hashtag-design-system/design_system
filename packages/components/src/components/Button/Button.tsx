import React from "react";
import { useClassnames, useDisabled } from "../../utils/hooks";

// const variants = {
//   checked: { pathLength: 1 },
//   initial: { pathLength: 0 },
// };

export const ButtonTypes = ["primary", "secondary", "danger"] as const;
export type ButtonType = typeof ButtonTypes[number];
const ButtonStates = ["default", "disabled", "focus-visible", "hover"] as const;
export type ButtonState = typeof ButtonStates[number];

export type Props = {
  type?: ButtonType;
  state?: ButtonState;
  block?: boolean;
  pill?: boolean;
};

export type FProps = Props & Omit<React.ComponentPropsWithRef<"button">, "type">;

const Button: React.FC<FProps> = ({ type = "primary", state = "default", block = false, pill = false, children, ...props }) => {
  let [classNames, rest] = useClassnames(
    `btn btn-${type} ${block ? "block" : ""} ${pill ? "pill" : ""} btn-default-font shadow__form-2`,
    props,
    { stateToRemove: { state } }
  );
  const isDisabled = useDisabled(props, state);

  // Animation state
  // const pathLength = useMotionValue(0);

  return (
    <button className={classNames} disabled={isDisabled} aria-disabled={isDisabled} data-testid="btn" {...rest}>
      {children}
    </button>
    /* {state === "loading" && (
        <motion.svg className="btn__loading-icon" width={78} height={41} fill="transparent">
          <motion.path
            d="M71.15 1H6a5 5 0 00-5 5v28a5 5 0 005 5h65.15a5 5 0 005-5V6a5 5 0 00-5-5z"
            strokeWidth={4}
            className="btn__loading-icon__path"
            variants={tickVariants}
            animate="checked"
            transition={{ duration: 5 }}
            style={{ pathLength }}
          />
        </motion.svg>
      )} */
    // </div>
  );
};

Button.displayName = "Button";

export default Button;
