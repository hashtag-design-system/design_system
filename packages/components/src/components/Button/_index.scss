// Primary button
.btn {
  @extend .flex-row-center-center;
  flex: none;
  align-self: flex-start;
  position: relative;
  white-space: nowrap;
  padding: em(8) em(16);
  width: auto;
  align-items: center;
  border-radius: 4px;
  border: none;
  outline: none;
  overflow: hidden;
  position: relative;
  text-align: center;
  height: em(36);
  appearance: none;
  transition: 0.2s cubic-bezier(0.3, 0, 0.5, 1);

  @include focus-visible-classNames() {
    @include accessibility-5dp;
  }

  // Pill button
  &.pill {
    border-radius: 24px !important;
  }

  &__loading-icon {
    position: relative;
    top: -3px;
    left: -76px;
    align-self: center;

    &__path {
      stroke: lighten($primary, 20);
    }
  }
}

.btn-primary {
  background-color: $primary;
  color: $grey-1;

  @include hover-classNames() {
    background-color: darken($primary, 10);
    border-color: darken($primary, 10);
    cursor: pointer;
  }

  @include disabled-classNames {
    @include disabled-state;
    background-color: $grey-7;
    border-color: $grey-7;
    color: $grey-10;
  }
}

// Secondary button
.btn-secondary {
  background-color: $grey-1;
  border: 2px solid $primary;
  box-sizing: border-box;

  @include hover-classNames() {
    background-color: $primary;
    cursor: pointer;
    color: $grey-1;
    border-color: $primary;
  }

  @include disabled-classNames {
    @include disabled-state;
    background-color: inherit;
    border-color: $grey-8;
    color: $grey-8;
  }
}

// Danger button
.btn-danger {
  color: $grey-1;
  background-color: $red-6;
  border-color: $red-6;

  @include hover-classNames() {
    background-color: $red-7;
    border-color: $red-7;
    cursor: pointer;
    color: $grey-1;
  }

  @include disabled-classNames {
    @include disabled-state;
    background-color: inherit;
    border-color: $red-6;
    color: $red-6;
  }

  @include focus-visible-classNames() {
    background-color: $red-6;
    border-color: $red-6;
  }
}
