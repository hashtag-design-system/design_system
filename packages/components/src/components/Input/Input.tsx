import React from "react";
import { IconPropType } from "../../typings";
import { InputContextProvider } from "../../utils/contexts";
import { generateInputId } from "../../utils/hooks/useInputId";
import { InputAutocompleteType } from "../__helpers__";
import Digit from "./Digit";
import DigitSequence from "./DigitSequence";
import IncrDcr from "./IncrDcr";
import { InputBaseFProps } from "./index";
import Multiline from "./Multiline";
import Number from "./Number";
import Password from "./Password";
import { Base, FieldContainer, InputContainer } from "./__helpers__";

const InputBaseStates = ["default", "focus", "disabled"] as const;
export type InputBaseState = typeof InputBaseStates[number];

export const InputStates = [...InputBaseStates, "hover", "error", "success"] as const;
export type InputState = typeof InputStates[number];

export type InputHelpTextType = {
  value: string;
  icon?: IconPropType;
  error?: boolean;
  transparent?: boolean;
};

type State = {
  id: FProps["id"];
  value: FProps["value"];
};

export type Props = {
  defaultValue?: React.ReactText;
  helptext?: InputHelpTextType;
  secondhelptext?: InputHelpTextType;
  characterLimit?: boolean;
  floatingplaceholder?: boolean | { now: boolean };
  label?: string;
  allowClear?: boolean;
  autoComplete?: InputAutocompleteType;
  optional?: boolean;
  forwardref?: React.ComponentPropsWithRef<"input">["ref"];
  passwordboxes?: React.ReactNode;
};

export type FProps = Props & InputBaseFProps<InputState>;

export type SBProps = Omit<Props, "passwordboxes"> &
  Pick<FProps, "placeholder" | "state" | "allowClear" | "suffix" | "prefix" | "type" | "maxLength">;

export default class Input extends React.Component<FProps, State> {
  static Base: typeof Base;

  static Password: typeof Password;

  static Multiline: typeof Multiline;

  static Digit: typeof Digit;

  static DigitSequence: typeof DigitSequence;

  static Number: typeof Number;

  static IncrDcr: typeof IncrDcr;

  static displayName = "Input";

  state: State = {
    id: this.props.id || "",
    value: this.props.value || this.props.defaultValue || "",
  };

  componentDidMount = () => {
    if (!this.state.id) {
      this.setState(prevState => ({
        ...prevState,
        id: generateInputId(),
      }));
    }
  };

  handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    // `overrideOnChange` Prop is being handle by the <Base /> helper component
    const { onChange } = this.props;

    this.setState({ value: e.target.value });

    if (onChange) {
      onChange(e);
    }
  };

  render() {
    const { id, value } = this.state;

    const { value: propsValue, overrideOnChange } = this.props;
    // console.log(id);

    return (
      <InputContextProvider
        value={{
          ...this.props,
          id,
          value: overrideOnChange && propsValue !== undefined ? propsValue : value,
          onChange: overrideOnChange ? this.props.onChange : this.handleChange,
          overrideOnChange,
        }}
      >
        <InputContainer>
          <FieldContainer />
        </InputContainer>
      </InputContextProvider>
    );
  }
}

// 6 (+1 <Base /> helper) sub-components in total
// ! Always add sub-components after declaration inside the component
Input.Base = Base;
Input.Password = Password;
Input.Multiline = Multiline;
Input.Digit = Digit;
Input.DigitSequence = DigitSequence;
Input.Number = Number;
Input.IncrDcr = IncrDcr;
