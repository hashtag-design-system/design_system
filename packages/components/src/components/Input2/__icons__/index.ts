export * from "./AddIcon";
export * from "./ClosedEye";
export * from "./OpenEye";
export * from "./SubtractIcon";
