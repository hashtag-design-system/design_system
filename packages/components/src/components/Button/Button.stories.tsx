import { Meta, Story } from "@storybook/react";
import Button, { ButtonFProps } from "./index";

export default {
  title: "Button",
  component: Button,
  args: {
    children: "Button",
  },
} as Meta;

const Template: Story<ButtonFProps> = ({ children, ...args }) => (
  <Button onClick={() => console.log("clicked")} {...args}>
    {children}
  </Button>
);

export const Default = Template.bind({});

export const Secondary = Template.bind({});
Secondary.args = {
  variant: "secondary",
};

export const Danger = Template.bind({});
Danger.args = {
  variant: "danger",
};
