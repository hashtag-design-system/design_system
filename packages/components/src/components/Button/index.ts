import Button from "./Button";

export type { ButtonState, ButtonType, Props as ButtonProps, FProps as ButtonFProps } from "./Button";
export { ButtonTypes } from "./Button";

export default Button;
