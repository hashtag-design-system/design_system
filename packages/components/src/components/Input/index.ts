import Input from "./Input";

export type { Props as InputProps } from "./Input";
export type {
  GroupProps as InputNumberGroupProps,
  NumberProps as InputNumberProps,
  StepperProps as InputNumberStepperProps,
} from "./Number";
export type { Props as InputTextareaProps } from "./TextArea";

export default Input;
