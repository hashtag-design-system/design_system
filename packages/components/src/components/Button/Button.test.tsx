import { render, screen } from "@testing-library/react";
import { OpenEye } from "../Input/__icons__";
import Button, { ButtonTypes } from "./index";

describe("<Button />", () => {
  test("default behaviour", () => {
    render(<Button />);
    const btn = screen.getByTestId("btn");

    expect(btn).toBeVisible();
    expect(btn).toHaveAttribute("class");
    expect(btn).not.toHaveAttribute("disabled");
    expect(btn).not.toBeDisabled();
    expect(btn).toHaveTextContent("");
    expect(btn.children).toHaveLength(0);
    expect(btn).toMatchSnapshot();
  });
  test.each(ButtonTypes)("btn type", type => {
    render(<Button type={type} />);

    expect(screen.getByTestId("btn").className).toContain(type);
  });
  test("block={true}", () => {
    render(<Button block />);

    expect(screen.getByTestId("btn")).toHaveClass("block");
  });
  test("pill={true}", () => {
    render(<Button pill />);

    expect(screen.getByTestId("btn")).toHaveClass("pill");
  });
  test("with children", () => {
    render(<Button>Button</Button>);
    const btn = screen.getByTestId("btn");

    expect(btn).toBeVisible();
    expect(btn).toHaveTextContent("Button");
    // Due to text, and not HTML element
    expect(btn.children.length).toBe(0);
  });
  test("with children of `React.ReactNode` (component - icon)", () => {
    render(
      <Button>
        Button
        <OpenEye />
      </Button>
    );
    const btn = screen.getByTestId("btn");

    expect(btn).toBeVisible();
    expect(btn).toHaveTextContent("Button");
    expect(btn.children.length).toBe(1);
    expect(btn).toContainHTML("<svg");
    expect(screen.getByTestId("icon")).toBeVisible();
  });
  describe("disabled state", () => {
    test('state="disabled"', () => {
      render(<Button state="disabled" />);
      const btn = screen.getByTestId("btn");

      expect(btn).toHaveClass("disabled");
      expect(btn).toBeDisabled();
      expect(btn).toHaveAttribute("aria-disabled", "true");
    });
    test('aria-disabled="true"', () => {
      render(<Button aria-disabled="true" />);
      const btn = screen.getByTestId("btn");

      expect(btn).toBeDisabled();
      expect(btn).toHaveAttribute("aria-disabled", "true");
    });
    test("disabled={true}", () => {
      render(<Button disabled />);
      const btn = screen.getByTestId("btn");

      expect(btn).toBeDisabled();
      expect(btn).toHaveAttribute("aria-disabled", "true");
    });
  });
});
