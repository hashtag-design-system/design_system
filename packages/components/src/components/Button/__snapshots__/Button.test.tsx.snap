// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> default behaviour 1`] = `
<button
  class="btn btn-primary   btn-default-font shadow__form-2"
  data-testid="btn"
/>
`;
